server/datastore/mysql/apple_mdm.go

```diff
@@ -2159,11 +2143,16 @@ func (ds *Datastore) bulkSetPendingMDMAppleHostProfilesDB(
 	profileIntersection := apple_mdm.NewProfileBimap()
 	profileIntersection.IntersectByIdentifierAndHostUUID(wantedProfiles, currentProfiles)
 
-	// start by deleting any that are already in the desired state
+	// Start by cleaning up profiles that have matching identifiers but different UUIDs (one way this can
+	// happen is a profile is deleted and then same profile is re-uploaded). We need to remove the
+	// old row here because we upsert below based on the UUID so the mis-match would lead to an orphan row
+	// (i.e. it wouldn't trigger the ON DUPLICATE KEY UPDATE).
```

Comment:
Note that we're relying on the intersection to detect these cases, but there's no guarantee that the intersection hasn't quietly discarded some cases that we would otherwise want to clean up.